src/store/compression_lz4_block.rs

```diff
@@ -0,0 +1,39 @@
+use std::io::{self};
+
+use core::convert::TryInto;
+use lz4_flex::{compress_into, decompress_into};
+/// Name of the compression scheme used in the doc store.
+///
+/// This name is appended to the version string of tantivy.
+pub const COMPRESSION: &str = "lz4_block";
+
+pub fn compress(uncompressed: &[u8], compressed: &mut Vec<u8>) -> io::Result<()> {
+    compressed.clear();
+
+    compressed.extend_from_slice(&[0, 0, 0, 0]);
+    compress_into(uncompressed, compressed);
+    let size = uncompressed.len() as u32;
+    compressed[0] = size as u8;
+    compressed[1] = (size >> 8) as u8;
+    compressed[2] = (size >> 16) as u8;
+    compressed[3] = (size >> 24) as u8;
+    Ok(())
+}
+
+pub fn decompress(compressed: &[u8], decompressed: &mut Vec<u8>) -> io::Result<()> {
+    decompressed.clear();
+    //next lz4_flex version will support slice as input parameter.
+    //this will make the usage much less ugly
+    let size = compressed.get(..4).ok_or(io::ErrorKind::InvalidData)?;
```

Comment:
no shadowing please. I know it is allowed by rust and clippy but it sucks. 
Here the name does not match the type either.

num_uncompress_bytes_arr = ...

Reply:
Generally I don't agree, I've seen a lot of bugs in my last project because there was no shadowing allowed (most notably outer db connection used in inner scope)

Reply:
so the shadowing was used to drop stuff?

Reply:
Interesting... I'll be happy to reconsider if such a case comes up shadowing in a scope to precisely "hide" something from outerscope.

Reply:
No, it was used as shadowing for a inner scope, roughly like this

```
let my db_con = get_new_connection()
// do stuff with db_con
{
    let my db_con = get_new_connection()
    // do stuff in separate transaction with inner db con
    // if outer db connection was used sometimes, really bad things did happen like a transaction deadlock
    db_con.commit();
}
// continue on outer connection
db_con.commit();
```

Reply:
Yes sorry I understood after and edited my message accordingly.